activedirectory/azure-ad-integration-spring-boot-autoconfigure/src/main/java/com/microsoft/azure/autoconfigure/adintegration/AzureADJwtTokenFilter.java

```diff
@@ -0,0 +1,118 @@
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.azure.autoconfigure.adintegration;
+
+import com.microsoft.aad.adal4j.AuthenticationContext;
+import com.microsoft.aad.adal4j.AuthenticationResult;
+import com.microsoft.aad.adal4j.ClientAssertion;
+import com.microsoft.aad.adal4j.ClientCredential;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.naming.ServiceUnavailableException;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class AzureADJwtTokenFilter extends OncePerRequestFilter {
+    private static final Logger log = LoggerFactory.getLogger(AzureADJwtTokenFilter.class);
+
+    private static final String TOKEN_HEADER = "Authorization";
+    private static final String TOKEN_TYPE = "Bearer ";
+
+    private AzureADJwtFilterProperties aadJwtFilterProp;
+
+    public AzureADJwtTokenFilter(AzureADJwtFilterProperties aadJwtFilterProp) {
+        this.aadJwtFilterProp = aadJwtFilterProp;
+    }
+
+    private AuthenticationResult acquireTokenForGraphApi(
+            String tokenEncoded,
+            String tenantId) throws Throwable {
+        ClientCredential credential = new ClientCredential(aadJwtFilterProp.getClientId(), aadJwtFilterProp.getClientSecret());
+        ClientAssertion assertion = new ClientAssertion(tokenEncoded);
+
+        AuthenticationResult result = null;
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
+            AuthenticationContext context = new AuthenticationContext(
+                    aadJwtFilterProp.getAadSignInUri()+ tenantId + "/",
+                    true,
+                    service);
+            Future<AuthenticationResult> future = context
+                    .acquireToken(aadJwtFilterProp.getAadGraphAPIUri(), assertion, credential, null);
+            result = future.get();
+        } catch (ExecutionException e) {
+            throw e.getCause();
+        } finally {
+            service.shutdown();
+        }
+
+        if (result == null) {
+            throw new ServiceUnavailableException("authentication result was null");
```

Comment:
Consider a more helpful message here, ie Unable to acquire OBO token for <GraphUri>